File: Part 2 - Regression/Multiple Linear Regression/multipler_linear_regression_P_valor.py
# -*- coding: utf-8 -*-
"""
Created on Thu Mar 30 19:16:37 2023

@author: josue
"""

# Regresion lineal multiple
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split
#from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression

dataset = pd.read_csv('50_Startups.csv')

# matriz. Obteniendo valores del data frame [filas,columnas]
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

# Codificar datos categoricos
ct = ColumnTransformer(
    [('one_hot_encoder', OneHotEncoder(categories='auto'), [3])],
    remainder='passthrough'
)
#NumPy con las columnas dummy adicionales creadas
X = np.array(ct.fit_transform(X), dtype=float)

#Evitar la trampa de las variables dummy.Podemos eliminar cualquiera
X = X[:, 1:]

#Dividir data set entre conjunto de entranamiento y testing
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size = 0.2, random_state=0)

# sc_X = StandardScaler()
# X_train = sc_X.fit_transform(X_train)
# X_test = sc_X.transform(X_test)

# Ajustar el modelo de regresion lineal multiple con el conjunto de entrenamiento
regression = LinearRegression()
regression.fit(X_train,y_train)#Teoricamente esto es all in

# Prediccion de los resultados en el conjunto de testing
y_pred = regression.predict(X_test)

# Construir el modelo optimo de RLM utilizando Eliminacion hacia atras
import statsmodels.api as sm
#Agregando termino independiente. Para calcular P valor
X = np.append(arr = np.ones((50,1)).astype(int), values = X, axis = 1)#0 fila 1 columna, al inicio 1's
def backwardElimination(x, sl):    
    numVars = len(x[0])    
    for i in range(0, numVars):        
        regressor_OLS = sm.OLS(y, x.tolist()).fit()        
        maxVar = max(regressor_OLS.pvalues).astype(float)        
        if maxVar > sl:            
            for j in range(0, numVars - i):                
                if (regressor_OLS.pvalues[j].astype(float) == maxVar):                    
                    x = np.delete(x, j, 1)    
    print(regressor_OLS.summary())   
    return x 
 
SL = 0.05
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
X_Modeled = backwardElimination(X_opt, SL)

File: Part 2 - Regression/Multiple Linear Regression/multipler_linear_regression_P_valor.R
# Regresion Lineal Multiple 

#importando dataset
dataset = read.csv('50_Startups.csv')


# Codificar datos categoricos
dataset$State = factor(dataset$State,
                       levels = c("New York","California","Florida"),
                       labels = c(1,2,3))

#Dividir data set entre conjunto de entranamiento y testing
library(caTools)
set.seed(123) 
split = sample.split(dataset$Profit,SplitRatio = 0.8)
training_set = subset(dataset,split==TRUE)
testing_set = subset(dataset,split==FALSE)

#Ajustar el modelo de Regresion Lineal Multiple con el Conjunto de Entranmiento
#. todas las otras
regression = lm(formula = Profit ~ ., data=training_set)
#P valor = Pr(>|t|)
summary(regression)

# Predecir los resultados con el conjunto de testing
y_pred = predict(regression, newdata = testing_set)

# Construir un modelo optimo con la Eliminacion hacia atrasq
backwardElimination <- function(x, sl) {
  numVars = length(x)
  for (i in c(1:numVars)){
    regressor = lm(formula = Profit ~ ., data = x)
    maxVar = max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
    if (maxVar > sl){
      j = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
      x = x[, -j]
    }
    numVars = numVars - 1
  }
  return(summary(regressor))
}

SL = 0.05
dataset = dataset[, c(1,2,3,4,5)]
backwardElimination(training_set, SL)


